narwhal/primary/src/state_handler.rs

```diff
@@ -185,16 +256,194 @@ impl RandomnessState {
+                        error!("random beacon: unable to write VSS key to output: {e:?}")
+                    }
+                    let num_shares = output.shares.as_ref().map_or(0, |shares| shares.len());
                     self.dkg_output = Some(output);
+                    info!("random beacon: DKG complete with {num_shares} shares for this node");
+                }
+                Err(fastcrypto::error::FastCryptoError::NotEnoughInputs) => (), // wait for more input
+                Err(e) => error!("random beacon: error while processing DKG Confirmations: {e:?}"),
+            }
+            // Begin randomness generation.
+            if self.dkg_output.is_some() {
+                info!("random beacon: start randomness generation");
+                self.send_partial_signatures().await;
+            }
+        }
+    }
+
+    async fn update_narwhal_round(&mut self, round: Round) {
+        self.narhwal_round = round;
+        // Re-send partial signatures to new leader, in case the last one failed.
+        if self.dkg_output.is_some() && (self.last_narwhal_round_sent <= round) {
+            self.send_partial_signatures().await;
+        }
+    }
+
+    async fn update_randomness_round(&mut self, round: RandomnessRound) {
+        if round <= self.randomness_round {
+            // Don't go backwards.
+            return;
+        }
+        self.randomness_round = round;
+        self.partial_sigs
+            .retain(|&(round, _), _| round >= self.randomness_round);
+        self.send_partial_signatures().await;
+    }
+
+    async fn send_partial_signatures(&mut self) {
+        let Some(dkg_output) = &self.dkg_output else {
+            error!("random beacon: called send_partial_signatures before DKG completed");
+            return;
+        };
+        let shares = match &dkg_output.shares {
+            Some(shares) => shares,
+            None => return, // can't participate in randomness generation without shares
+        };
+        debug!(
+            "random beacon: sending partial signatures for round {}",
+            self.randomness_round
+        );
+        if self.cached_sigs.is_none()
+            || self.cached_sigs.as_ref().unwrap().0 != self.randomness_round
+        {
+            self.cached_sigs = Some((
+                self.randomness_round,
+                ThresholdBls12381MinSig::partial_sign_batch(
+                    shares,
+                    &self.randomness_round.signature_message(),
+                ),
+            ));
+        }
+        let sigs = self.cached_sigs.as_ref().unwrap().1.clone();
+
+        // To compute next leader round, add two to even round, and one to odd round.
+        let next_leader_narwhal_round = (self.narhwal_round + 2) & !1;
+        self.last_narwhal_round_sent = next_leader_narwhal_round;
+
+        let leader = self.leader_schedule.leader(next_leader_narwhal_round);
+        if self.authority_id == leader.id() {
```

Comment:
I wanted to go through the same receive_partial_sigs flow to avoid any possibility later for bugs where logic in that fn needs to be duplicated here but isn't. I'll grant it's a bit less efficient this way